File: LeetCode/66PlusOne/PlusOne.py
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        # Single Pass -> Time:O(n), Space:O(n)
        s, carry = 0, 0
        for i in range(len(digits)-1, -1, -1):
            s = digits[i]
            if i == len(digits)-1:
                s += 1
            s += carry
            carry = s//10
            digits[i] = s%10
        if carry:
            digits.insert(0,1)
        return digits

        # Pythonic -> Time:O(n), Space:O(n)
        return [int(i) for i in str(int("".join(map(lambda x: str(x), digits)))+1)]

File: LeetCode/66PlusOne/PlusOne.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        // Single Pass -> Time:O(n), Space:O(1)
        int sum, carry = 0;
        for (int i = digits.size()-1; i > -1; i--) {
            sum = digits[i];
            if (i == digits.size()-1) {
                sum++;
            }
            sum += carry;
            digits[i] = sum%10;
            carry = sum/10;
        }
        if (carry) {
            digits.insert(digits.begin(),1);
        }
        return digits;
    }
};
